### PR TITLE
Fix #5838: update Cubical reductions according to Simon Huber

### LINKED ISSUE #5838: Proof of ⊥ with --safe --cubical
I stumbled upon a proof of ⊥ based on computations over univalence. In the example, I am using the cubical library.

I really don't understand anything more about what is going wrong. Is there a bug in transp on Glue when φ is not i0, maybe?

```agda
{-# OPTIONS --safe --cubical #-}

module LocalGlobalParadox where

open import Cubical.Foundations.Prelude
open import Cubical.Foundations.Equiv
open import Cubical.Foundations.Univalence
open import Cubical.HITs.S1
open import Cubical.Data.Empty
open import Cubical.Data.Nat
open import Cubical.Data.Int

-- Transposition of 2-loops
transpose : ∀ {ℓ} {A : Type ℓ} {x y z w : A} {p : x ≡ y} {q : y ≡ z} {r : x ≡ w} {s : w ≡ z} → PathP (λ i → p i ≡ s i) r q → PathP (λ i → r i ≡ q i) p s
transpose sq i j = sq j i

-- Transposition of 2-loops is equal to inversion (sym)
transpose-sym : ∀ {ℓ} {A : Type ℓ} {x : A} (p : Path (x ≡ x) refl refl) →
  Path (Path (x ≡ x) refl refl)
       (transpose p)
       (sym p)
transpose-sym p i j k =
  hcomp (λ l → λ { (i = i0) → p k j
                 ; (i = i1) → p (~ j) k
                 ; (j = i0) → p (i ∨ k ∨ l) (i ∧ k)
                 ; (j = i1) → p (~ i ∧ k ∧ ~ l) (~ i ∨ k)
                 ; (k = i0) → p (i ∧ ~ j ∧ ~ l) (~ i ∧ j)
                 ; (k = i1) → p (~ i ∨ ~ j ∨ l) (i ∨ j)
                 })
        (p ((~ j ∧ k) ∨ (i ∧ ~ j) ∨ (~ i ∧ k)) ((j ∧ k) ∨ (~ i ∧ j) ∨ (i ∧ k)))

-- Candidate implementation of "local-global looping principle" as
-- Nicolai Kraus called it. I am not certain if these are really
-- correct, but they demonstrate the bug.
global : ∀ {ℓ} {A : Type ℓ} → ((x : A) → x ≡ x) → Path (A ≡ A) refl refl
global {A = A} h i j =
  Glue A (λ { (i = i0) → A , idEquiv A
            ; (i = i1) → A , idEquiv A
            ; (j = i0) → A , equivEq {e = idEquiv A} {f = idEquiv A} (λ k x → h x k) i
            ; (j = i1) → A , idEquiv A
            })

-- This seems OK:
local : ∀ {ℓ} {A : Type ℓ} → Path (A ≡ A) refl refl → ((x : A) → x ≡ x)
local {A = A} h x = sym (transportRefl x) ∙∙ (λ i → transp (λ j → h i j) i0 x) ∙∙ transportRefl x

-- This one seems problematic:
local' : ∀ {ℓ} {A : Type ℓ} → Path (A ≡ A) refl refl → ((x : A) → x ≡ x)
local' h x i = transp (λ j → h i j) (i ∨ ~ i) x



-- Now a specific example, the circle S¹

-- 𝔾 is supposed to be equivalent to ℤ
𝔾 : Type₁
𝔾 = Path (Path Type₀ S¹ S¹) refl refl

𝟙 : 𝔾
𝟙 = global rotLoop

-- by above, these are supposed to be equal
*𝔾 -𝔾 : 𝔾 → 𝔾
*𝔾 = transpose
-𝔾 = sym

-- seemingly correct conversion to ints
toℤ : 𝔾 → ℤ
toℤ n = winding (local n base)

_ : toℤ (*𝔾 𝟙) ≡ toℤ (-𝔾 𝟙)
_ = refl {x = -1}

-- problematic conversion to ints
problem : 𝔾 → ℤ
problem n = winding (local' n base)

-- it doesn't respect transpose-sym!
_ : problem (*𝔾 𝟙) ≡ 0 -- ???
_ = refl

_ : problem (-𝔾 𝟙) ≡ -1
_ = refl

-- so we get a contradiction
bad : 0 ≡ -1
bad = cong problem (transpose-sym (global rotLoop))

boom : ⊥
boom = transport (λ i → iszero (bad i)) tt
  where
  iszero : ℤ → Type₀
  iszero (pos zero) = Unit
  iszero _ = ⊥
```

### PR DESCRIPTION
**Pull request by andreasabel**
Fix #5838: update Cubical reductions according to Simon Huber.

See discussion at #5755.

### CODE REVIEW

**Comment by mortberg** [comment]
Nice! Does this break anything in the cubical library?

Like I wrote in the other thread we most likely need an analogous fix to transp for HCompU as that code is essentially just an unfolded and optimized version of transp for Glue in the special case of lines in U.

**Comment by andreasabel** [comment]
> Nice! Does this break anything in the cubical library?

So, the cubical-library test went through, both locally and on CI.  But I did not make any further investigation, e.g. whether there are new commits after the current pin that should also be tested.

**Comment by andreasabel** [comment]
@tomjack @mortberg @sattlerc : I modified `compHCompU DoTransp` in analogy to @simhu's suggested fix, please review!

**Comment by tomjack** [comment]
Excellent, thanks!

I tried to make a similar change in a different way, and the result on the 5838b example was that Agda rapidly consumed the entire heap. I don't understand why, but, your change looks like what I wanted to write, and actually seems to work. :smile:

**Comment by andreasabel** [comment]
@tomjack wrote:
> but, your change looks like what I wanted to write,

No coincidence, I followed your original patch...

File: src/full/Agda/TypeChecking/Primitive/Cubical.hs [hunk 1]
```diff
@@ -554,21 +556,33 @@ unglueTranspGlue psi u0 (IsFam (la, lb, bA, phi, bT, e)) = do
...
+        --
+        -- See: https://github.com/agda/agda/issues/5755#issuecomment-1043797776
+
+        -- unglue_u0 i = unglue la[i/i] lb[i/i] bA[i/i] phi[i/i] bT[i/i] e[i/e] u0
+        let unglue_u0 i =
+              foldl (<#>) (pure tunglue) (map (<@> i) [la, lb, bA, phi, bT, e]) <@> u0
+
         view <- intervalView'
 
         -- phi1 <- view <$> (reduce =<< (phi <@> pure io))
```
Comment (mortberg) [inline src/full/Agda/TypeChecking/Primitive/Cubical.hs]: What is this commented code?

**Review by mortberg — APPROVED** [review]
The changes look good to me! 

Does the "cubical-library test" check the latest master branch of agda/cubical? If so it's great news that we don't rely on this bug anywhere in the whole library. I was a little nervous...

File: src/full/Agda/TypeChecking/Primitive/Cubical.hs (hunk 1 above)
Comment (andreasabel) [reply src/full/Agda/TypeChecking/Primitive/Cubical.hs]: This is just a comment, explaining the line below it.  But I see there is a bug in it: `e[i/e]` should be `e[i/i]`.

**Review by sattlerc — APPROVED** [review]
Looks good.

Someone should refactor this part of the code base so that the three instances become one.

File: src/full/Agda/TypeChecking/Primitive/Cubical.hs (hunk 1 above)
Comment (andreasabel) [reply src/full/Agda/TypeChecking/Primitive/Cubical.hs]: If you meant `-- phi1 <- view <$> (reduce =<< (phi <@> pure io))`, then I don't know, this was there before.

Comment (andreasabel) [reply src/full/Agda/TypeChecking/Primitive/Cubical.hs]: Fixed the comments.

Comment (mortberg) [reply src/full/Agda/TypeChecking/Primitive/Cubical.hs]: Great! And yeah, I meant the cryptic commented old code

**Comment by andreasabel** [comment]
@mortberg wrote:
> Does the "cubical-library test" check the latest master branch of agda/cubical? If so it's great news that we don't rely on this bug anywhere in the whole library.

If you mean `make cubical-test` which is run during the Agda CI, __no__, it checks the version pinned by the `cubical` submodule.
https://github.com/agda/agda/blob/46bc3773540ae65d750678c17e439f561e55f2c8/src/github/workflows/test.yml#L325-L326
https://github.com/agda/agda/blob/46bc3773540ae65d750678c17e439f561e55f2c8/Makefile#L523-L527
https://github.com/agda/cubical/blob/d29db7f5e6cfe5ddfa0d2c3b9336cf360ba99636/GNUmakefile#L12

This is because new additions to `cubical` might not work with the latest Agda `master`; the CI for `cubical` uses a release of Agda (2.6.2 atm, afaics).  

__(*)__ To have `make cubical-test` catch up with the latest cubical library, one runs:
```
make fast-forward-cubical
```
and then (after successful `make install-bin && make cubical-test`---the former to build the latest Agda) pushes the new submodule pin `cubical` to `master`.

I'll try this PR with the latest `cubical`. .. 
However, it would be good if you from time to time did __(*)__ to test whether Agda and `cubical` are still coherent.

**Comment by andreasabel** [comment]
> I'll try this PR with the latest `cubical`. .. However, it would be good if you from time to time did **(*)** to test whether Agda and `cubical` are still coherent.

Turns out that the latest `cubical` does not work with the latest Agda:
- https://github.com/agda/cubical/issues/743

I'll try to pick this PR onto Agda 2.6.2.2-rc2 and see whether this works with the latest cubical.

**Comment by mortberg** [comment]
> > I'll try this PR with the latest `cubical`. .. However, it would be good if you from time to time did **(*)** to test whether Agda and `cubical` are still coherent.
> 
> Turns out that the latest `cubical` does not work with the latest Agda:
> 
>     * [Reflection API changing in Agda 2.6.3 cubical#743](https://github.com/agda/cubical/issues/743)
> 
> 
> I'll try to pick this PR onto Agda 2.6.2.2-rc2 and see whether this works with the latest cubical.

Great! Please let me know if it worked.

**Comment by andreasabel** [comment]
> Great! Please let me know if it worked.

Looks like it worked:
https://github.com/agda/agda/runs/5708828503?check_suite_focus=true#step:9:27

I'll include this PR in Agda 2.6.2.2-rc2 then.

**Pull request merged**